Switch to aws-sdk S3 client

sktt opened the pull request:
Hi there!

The knox s3 client seems to not be maintained very well so here's an example how we can do the same same thing using the official aws-sdk.

Let me know if you're interested in this and I'd clean it up, update readme, tests and whatnot

Cheers

sktt committed:
Switch to aws-sdk S3 client

sktt commented on index.js:
```diff
@@ -8,15 +8,16 @@ TODO
+	accessKeyId: process.env.S3_KEY || process.env.AWS_ACCESS_KEY_ID,
+	secretAccessKey: process.env.S3_SECRET || process.env.AWS_SECRET_ACCESS_KEY,
+	region: process.env.S3_REGION || process.env.AWS_REGION,
+	bucket: process.env.S3_BUCKET || 'us-standard',
```
Here allowing both aws default env var names and the previous env var names.. Should we go with just one?

sktt commented on index.js:
```diff
@@ -104,27 +93,24 @@ S3Adapter.prototype.uploadFile = function (file, callback) {
+		var params = {
+			ContentLength: file.size,
+			ContentType: file.mimetype,
+			Body: fs.readFileSync(localpath),
```
todo

sktt commented on index.js:
```diff
@@ -150,9 +136,7 @@ S3Adapter.prototype.uploadFile = function (file, callback) {
-	// Consider providing an option to use insecure http. I can't think of any
-	// sensible use case for plain http though. https should be used everywhere.
-	return this._knoxForFile(file).https(this._resolveFilename(file));
+	return 'https://' + (file.bucket || this.options.bucket) + '.s3.amazonaws.com' + this._resolveFilename(file);
```
maybe bad idea to have `.s3.amazonaws.com` hardcoded like this.

burgerboydaddy replied:
Please note in China S3 url's are different. As per  [http://docs.amazonaws.cn/en_us/aws/latest/userguide/s3.html](url)

> The endpoints for services in the Beijing Region are different from other global endpoints. For an endpoint in the Beijing Region, the Region value should be cn-north-1 and the domain should be amazonaws.com.cn. 

In general URL end point is `.s3.cn-north-1.amazonaws.com.cn`

sktt replied:
Thank you!

Maybe it would work to use `	return 'https://' + (file.bucket || this.options.bucket) + '.' + this.client.endpoint.host + this._resolveFilename(file);
`

sktt commented on index.js:
```diff
@@ -69,20 +72,6 @@ S3Adapter.SCHEMA_FIELD_DEFAULTS = {
-	} else {
-		return this.client;
-	}
-};
```
the S3 client can handle multiple bucket names. but anyway for this instance, we should be using only `this.options.bucket`. Should it ever happen that the `file.bucket` and `options.bucket` differ? wheN?

sktt committed:
Replace hardcoded host part with endpoint.host

sktt committed:
Use the `key` and `secret` options as current ver

sktt committed:
Bunch of edits

  - Remove unused code
  - Rename _resoveFilename to _resolveKey
  - Pass `params` on to s3 client

sktt committed:
Allow params to be a function

  - Creds to @steevsachs https://github.com/keystonejs/keystone-storage-adapter-s3/pull/18

sktt committed:
Throw when passing unsupported `headers` option

sktt committed:
2.0.0

sktt committed:
Bug: Key should not have a preceding slash

r3wt commented:
@JedWatson status on this?

michaek commented:
It seems this should probably be merged, even if it's not a perfect replacement for the knox implementation. Support for IAM roles seems pretty essential nowadays.

bflopez commented:
Is there another thread to follow for replacing Knox with aws-sdk? Researching KeystoneJS right now for a project and this is one of the things holding me back from starting it. Can't get S3 to work.

mikehazell commented:
closing this in favour of #35
